### ISSUE TITLE
Many deprecation warnings with FFmpeg 3.2 and template errors with latest antlr3

### ISSUE THREAD

**Issue by rbalint** [owntone/owntone-server#307]
The build log on Debian unstable shows the issues with fairly up to date dependencies:
https://buildd.debian.org/status/fetch.php?pkg=forked-daapd&arch=amd64&ver=24.2-1&stamp=1478814383

**Comment by ejurgensen** [comment]
I'm working on updating to the new ffmpeg and libav api's in this branch https://github.com/ejurgensen/forked-daapd/tree/ffmpeg_update1, and luckily it seems that these newer versions are much more aligned, meaning that I can remove a lot of the cross-compability stuff.

However, I still want to maintain compability with the older versions, and my intention is to do this by looking for avcodec_send_packet() and avcodec_parameters_alloc(). If they are not present, I will assume that it is an old version, and build with my old source files (now called transcode_legacy.c and artwork_legacy.c) instead of the new ones (transcode.c and artwork.c).

@sshambar, can you help me with the configure.ac magic required to do this? Note that if the new API is detected then almost all the old feature checks can be skipped (only requirement is to set HAVE_FFMPEG and HAVE_DECL_AVCODEC_PARAMETERS_ALLOC).

**Comment by sshambar** [comment]
@ejurgensen I actually rewrote transcode.c do exactly what you're talking about... but was working on getting video transcode working properly before I submitted it (it works now, but the muxing isn't correct for streaming yet...).  If you like, I can post my work on a branch, and you can check it out.

**Comment by sshambar** [comment]
OK, I've pushed my WIP on transcode.c to my fixwarns branch.  It doesn't use the send_packet etc apis yet as I've been setting the groundwork for minimal code change for the new apis while still working correctly on the old apis.  The changes should also work with video transcoding (ffmpeg only atm), but I'm still tweaking the mux parameters (fragmented keyframes) to correctly interleave for minimally delayed input/output.  Getting the dts/pts timing correct required lots of digging in the ffmpeg code to get correct for video, and I haven't had a chance to do the same in libav, so the video code doesn't yet compile with libav...(although it's easy to just ifdef out the framerate refs)

I reworked the code to make the filter handling work the same in ffmpeg and libav (more unified code).  Also, the code uses a private codec_ctx (rather than the stream one), and should work correctly on the old codec or new codecpar apis with only a few ifdefs.

It's NOT ready for a pull request (not libav ready), but the ffmpeg code works well in my tests so far... let me know what you think.
